### PR TITLE
Service environment

### PR DESCRIPTION
**Pull request by lieut-data**
#### Summary
Bring us closer to the [Single Build Artifact](https://docs.google.com/document/d/1LcHidviRESHZ6PgZs4gyGA_U1Npp-F2k6R0trMazqUY/edit) vision by embedding all "secrets" directly into the source instead of juggling a myriad of release pipelines that generate disparate binaries/images.

Different environments must be serviced differently, i.e. discrete Rudder keys between production and cloud. The `MM_SERVICEENVIRONMENT` controls this setting:

| MM_SERVICEENVIRONMENT | Description |
| --- | --- |
| production | The default if unset for production builds |
| test | Testing environment (e.g. `/cloud`)
| dev | Development environments, the default if unset for non-production builds |

#### Ticket Link
Fixes: https://mattermost.atlassian.net/browse/MM-52707

#### Release Note
```release-note
NONE
```

### CODE REVIEW

File: server/channels/utils/license.go
Comment (marianunez) [inline server/channels/utils/license.go:79]: What would be different for the cloud environment from the prod environment? It seems that in this PR it's always considered the same case as Default which seems to be the case for Rudder, Licenses and Stripe keys

File: server/Makefile
Comment (marianunez) [inline server/Makefile:51]: Should we automatically set the external environment env var here to `Dev` so it's defaulted for local dev?

File: server/channels/utils/license.go
Comment (lieut-data) [reply server/channels/utils/license.go:79]: Good question. Somehow, I assumed going in that we needed to distinguish cloud from self-managed in some way, given (I think) we have separate Docker images for each, but you're right that after these changes nothing obvious shook out.

cc @stylianosrigas, any idea what distinguishes these images from your side?

File: server/Makefile
Comment (lieut-data) [reply server/Makefile:51]: I can certainly do that, but I found that wasn't sufficient, since a non-Makefile invocation like `go test` wouldn't automatically get the setting applied. In `model.GetExternalServiceEnvironment`, I'm detecting the presence of a non-dev build number to try to assert with confidence that we're allowed to use the production external environment.

**Review by marianunez — COMMENTED** [review]
Thanks @jesse! Left some initial comments for a first high level pass, will come back for full review. 

`External Service Environment` is not as clear to me what it does so agree that we could consider some other alternatives. I don't have a suggestion yet but will think about it

File: server/channels/utils/license.go
Comment (phoinixgrr) [reply server/channels/utils/license.go:79]: Hey @stylianosrigas  @lieut-data 

I apologize for any inconvenience caused by my lack of context. If you could kindly provide some additional information, I would greatly appreciate it. I would like to understand exactly what information is needed here.

Comment (lieut-data) [reply server/channels/utils/license.go:79]: Hi @phoinixgrr! I'm helping wrap up work that @toninis, Nathaniel, @marianunez and I started to help us realize the delivery team's vision of a single build artifact that conforms to the 12 factor principle, instead of our current practice of a myriad of release pipelines that build various Docker images, e.g. `mattermost-enterprise-edition` for self-managed customers, `mattermost-cloud` for Cloud, and `mm-ee-test` for testing environments.

This PR makes it possible for a single binary to be any of the above with the appropriate environment variable, but after incorporating all necessary changes, we found no need to distinguish between the `mattermost-enterprise-edition` and the `mattermost-cloud`. Do you know why we build separate images in this case? Is there some other difference we need to account for in these application changes? Or is it simply that we publish builds from the `cloud` branch to a different image?

File: server/channels/app/server.go
Comment (lieut-data) [reply server/channels/app/server.go:425]: I'm game for that!

File: server/channels/utils/fileutils/fileutils_test.go
Comment (lieut-data) [reply server/channels/utils/fileutils/fileutils_test.go:105]: Yes, I pulled this out as a discrete commit fixing an issue that prevented running the related tests on MacOS: https://github.com/mattermost/mattermost-server/pull/23443/commits/b02973aa904e4a3b1a5ecb1b20ef5c75b6ccf54f.

File: server/channels/utils/license_test.go
Comment (lieut-data) [reply server/channels/utils/license_test.go:74]: Yes, this test isn't as comprehensive as I'd like. Ideally, I'd check in a valid production license and a valid test license and make sure they get accepted only with their compatible certificates, but it seemed like the pattern here was all "failure testing". Thoughts on adding such a sample license? Or maybe we already have them somewhere?

File: server/platform/services/telemetry/telemetry.go
Comment (lieut-data) [reply server/platform/services/telemetry/telemetry.go:171]: Agreed, will unset consistently for dev.

File: server/public/model/config.go
Comment (lieut-data) [reply server/public/model/config.go:2825]: Thank you, I changed this and forgot to revisit. Fixed!

**Comment by lieut-data** [comment]
@marianunez, thanks again for the feedback. To make this whole thing more robust, I reintroduced the go tags approach with production builds being explicitly tagged and triggering a different default service environment. Here's the matrix:

| MM_SERVICESETTINGS | -tags production | Effective Environment |
| ------------------------ | ----------------- | ---------------------- |
| unset or empty or unknown value | YES | enterprise |
| unset or empty or unknown value | NO | dev |
| set to X | * | X |

When using `go test` or most `Makefile` directives, there is no `-tags production` and so we get the default `dev` behaviour. When building via `make package`, the `-tags production` ensures the default of `enterprise`. The environment variable can always be used to override `MM_SERVICESETTINGS`, removing some of the previous magic around build number detection.

Furthermore, this now introduces a discrete `enterprise` service environment, adding to the `cloud`, `test`, and `dev`. It's still not super clear that we need this level of granularity, but it makes it very easy to get precisely the desired behaviour especially as we find the need to build on this functionality.

**Comment by lieut-data** [comment]
I've also filed https://mattermost.atlassian.net/browse/CLD-5691 to the SRE team to help export this environment variable in advance of our actually deploying.

File: webapp/boards/src/index.tsx
Comment (marianunez) [inline webapp/boards/src/index.tsx:523]: nit:
```suggestion
            case ServiceEnvironment.DEFAULT:
            case ServiceEnvironment.CLOUD:
                rudderKey = '1myWcDbTkIThnpPYyms7DKlmQWl'
                break
```

**Review by marianunez — APPROVED** [review]
Thanks @lieut-data! Minor non-blocking comments below

**Comment by marianunez** [comment]
> https://mattermost.atlassian.net/browse/CLD-5691



> @marianunez, thanks again for the feedback. To make this whole thing more robust, I reintroduced the go tags approach with production builds being explicitly tagged and triggering a different default service environment. Here's the matrix:
> 
> MM_SERVICEENVIRONMENT	-tags production	Effective Environment
> unset or empty or unknown value	YES	enterprise
> unset or empty or unknown value	NO	dev
> set to X	*	X
> When using `go test` or most `Makefile` directives, there is no `-tags production` and so we get the default `dev` behaviour. When building via `make package`, the `-tags production` ensures the default of `enterprise`. The environment variable can always be used to override `MM_SERVICESETTINGS`, removing some of the previous magic around build number detection.
> 
> Furthermore, this now introduces a discrete `enterprise` service environment, adding to the `cloud`, `test`, and `dev`. It's still not super clear that we need this level of granularity, but it makes it very easy to get precisely the desired behaviour especially as we find the need to build on this functionality.

Thanks @lieut-data! This hardens the local dev scenario much more but we still need to make sure we properly override for all testing scenarios.  There are several servers that are managed directly by the QA Team so heads up @saturninoabril that the env var may be needed so they keep working with the latest updated test licenses. 

I still don't know if we need the `cloud` environment distinction but doesn't hurt as-is.

**Comment by marianunez** [comment]
> Creating a new SpinWick test server using Mattermost Cloud.

@lieut-data setting up a test server that should come up with the dev environment

File: server/channels/utils/license_test.go
Comment (marianunez) [reply server/channels/utils/license_test.go]: For the sample license maybe we can do just a test license and validate that it passes for dev/test and fails others. This way we don't add a sample prod license in the repo. Let me know if this works for you and I can provide a payload

Comment (lieut-data) [reply server/channels/utils/license_test.go]: That would be great, thank you @marianunez!

**Comment by lieut-data** [comment]
Thank you for helping check, @marianunez. Can you confirm this part:

> ... should come up with the dev environment

Unless `MM_SERVICESETTINGS=dev` or it's a development build, I think it /should/ default to production. Were you thinking we should compile at this stage with a non-production build, and reserve the `-tags production` to the GitLab build environment?

**Comment by lieut-data** [comment]
Note that I've marked this as `Do Not Merge` as we require https://mattermost.atlassian.net/browse/CLD-5691 to be completed before we can safely deploy to test and cloud environments.

The more I evaluate the current set of `cloud`, `enterprise`, `test` and `dev`, I'm thinking of perhaps aligning everything to just `production` and `test`, effectively enabling the rudder development keys for local dev testing (if enabled, etc.). This would mean cloud production requires no export, and would behave just the same as self-managed (since we infer the other "cloud bits" from the license itself). Thoughts, @marianunez?

**Comment by marianunez** [comment]
> /e2e-test MM_ENV="MM_SERVICESETTINGS=test"

@lieut-data is the env var `MM_SERVICEENVIRONMENT`? This is what I see in the code but noticed how you called here your E2E tests are different. 
I changed the env var in the test server but doesn't seem to have changed the CWS url though 🤔 

```
{
    "ID": "won6rjtgrf8smcyidbcd4zdk6h",
    "OwnerID": "mattermost-server-pr-23443",
    "GroupID": "qyr58czntbyf3mitxy9gtutckr",
    "GroupSequence": 56,
    "Version": "9fd3999",
    "Image": "mattermostdevelopment/mm-ee-test",
    "Name": "mattermost-server-pr-23443",
    "Database": "aws-multitenant-rds-postgres-pgbouncer",
    "Filestore": "bifrost",
    "License": "",
    "MattermostEnv": {
        "MM_SERVICEENVIRONMENT": {
            "value": "test"
        }
    },
    ```

**Comment by lieut-data** [comment]
@marianunez, gah, thank you for spotting that @marianunez. I typoed once and then copy-pasted too many times before I found and fixed my original mistake. It should indeed be `MM_SERVICEENVIRONMENT`, and I'll be repeating the e2e tests shortly.

>  I changed the env var in the test server but doesn't seem to have changed the CWS url though 🤔

Is this after a pod restart?

**Comment by marianunez** [comment]
> > I changed the env var in the test server but doesn't seem to have changed the CWS url though 🤔
> 
> Is this after a pod restart?

That is what I believe given that the cloud plugin restarts the pods after a env var change AFAIK. 

I'm going to change the test server it to a CWS + Cloud test server so we can try the Stripe key as well.

**Comment by marianunez** [comment]
> The more I evaluate the current set of `cloud`, `enterprise`, `test` and `dev`, I'm thinking of perhaps aligning everything to just `production` and `test`, effectively enabling the rudder development keys for local dev testing (if enabled, etc.). This would mean cloud production requires no export, and would behave just the same as self-managed (since we infer the other "cloud bits" from the license itself).

This makes sense to me. Unless there is some other special consideration that I am not aware of in the cloud build pipelines, there doesn't seems any additional configuration on the application side would be needed specifically for `cloud`. +1 on merging it to just `production`

**Comment by lieut-data** [comment]
I took a look at changing the release pipeline to not fail on failing to `sed` replace the various `placeholder_*` keys, but the myriad of branches and repos were overwhelming. To sidestep this, I've re-introduced constant strings that I've verified pass the various `sed` commands and will allow the release pipelines to remain unmodified as they currently exist (but effectively get ignored). When we move to the new GitHub pipelines, we can rip out these deprecated placeholders.

**Comment by saturninoabril** [comment]
/e2e-test MM_ENV="MM_SERVICEENVIRONMENT=test" SERVER_TYPE=cloud

File: server/channels/utils/license.go
Comment (saturninoabril) [reply server/channels/utils/license.go]: (Chiming in)

> Or is it simply that we publish builds from the cloud branch to a different image?

This is what I thought too.

Since the introduction of cloud, in terms of testing, I'm not treating `mattermost-cloud` differently from `mattermost-enterprise-edition` (or their equivalent test images), except for loading cloud license, cloud-specific server config (together with test CWS test, etc) and cloud-specific env variables (api key, customer and installation IDs).

**Comment by lieut-data** [comment]
Thank you, @saturninoabril! What is the `SERVER_TYPE` parameter?

**Comment by saturninoabril** [comment]
You're welcome :)
Only `onprem` and `cloud`. Default is `onprem` if not provided.

**Comment by lieut-data** [comment]
Ah, cool! I guess this runs a different set of tests?

**Comment by saturninoabril** [comment]
Correct, running applicable cloud tests against cloud-like environment.

**Comment by saturninoabril** [comment]
I ran `mattermostdevelopment/mm-ee-test:92b5cc0` without setting up `MM_SERVICEENVIRONMENT` and I can see that `CloudSettings` config are using production values which I'd expect to be test values. I'm guessing the same issue with other prod keys.

**Comment by lieut-data** [comment]
@saturninoabril, the plan is to deprecate `mm-ee-test`, converging on a single binary /image that alerts defaults to production values. Thereafter, to get test semantics, the `MM_SERVICEENVIRONMENT` will be required.

To ease this transition, we could update the `mm-ee-test` Docker image to run the server with this environment variable already exported. Thoughts?

**Comment by saturninoabril** [comment]
I'd lean towards not to update `mm-ee-test` and should be fine to go straight to the end goal. I guess my clarification would be what's the typical image convention:
- for master, maybe no change, same as `mattermost/mattermost-enterprise-edition:master`?
- for PR build, maybe `mattermost/mattermost-enterprise-edition:<short_sha>` (ex. `mattermost/mattermost-enterprise-edition:92b5cc0`)?
- for RC, maybe `mattermost/mattermost-enterprise-edition:<rc_tag>`?

Would be nice if the conventions could be updated here if possible, or done before merging this PR.

**Comment by lieut-data** [comment]
Got it, thank you @saturninoabril! My expectation aligns with yours in regards to image conventions and tagging off `mattermost/mattermost-enterprise-edition`. However, I note that this change merely unlocks the next phase of the Delivery Team's plan that would achieve those tagging conventions: after merging this PR, all the existing images will continue to be built and used normally, with the exception that testing environments need to have `MM_SERVICEENVIRONMENT=test`.

I'm aware of the need to do that for `/cloud` managed instances, and those involved in helping test a PR, but I expect this will have the most impact on QA. Thoughts on how to ensure we transition seamlessly?

**Comment by nickmisasi** [comment]
> The more I evaluate the current set of `cloud`, `enterprise`, `test` and `dev`, I'm thinking of perhaps aligning everything to just `production` and `test`, effectively enabling the rudder development keys for local dev testing (if enabled, etc.)

@lieut-data I think the distinction between `cloud` and `enterprise` isn't necessary and agree just `production` and `test` is probably the way to go. 

I do want to call out that enabling Rudder in local development would probably have some implications on our invoices since Rudder charges based off of event volume. We've run into some problems with other 3rd party's where CI pipelines eat through allotments. cc @catalintomai @ifoukarakis any concerns here?

I've played around with what's here locally and things seem to line up the way I'd expect them to - this is going to be great. Thanks for spear-heading this Jesse!

**Review by nickmisasi — APPROVED** [review]

**Comment by catalintomai** [comment]
> > The more I evaluate the current set of `cloud`, `enterprise`, `test` and `dev`, I'm thinking of perhaps aligning everything to just `production` and `test`, effectively enabling the rudder development keys for local dev testing (if enabled, etc.)
> 
> @lieut-data I think the distinction between `cloud` and `enterprise` isn't necessary and agree just `production` and `test` is probably the way to go.
> 
> I do want to call out that enabling Rudder in local development would probably have some implications on our invoices since Rudder charges based off of event volume. We've run into some problems with other 3rd party's where CI pipelines eat through allotments. cc @catalintomai @ifoukarakis any concerns here?
> 
> I've played around with what's here locally and things seem to line up the way I'd expect them to - this is going to be great. Thanks for spear-heading this Jesse!

Thanks for tagging us to this PR @nickmisasi . Yes, I am not fully convinced that the `test` and `dev` conflation is the way to go - at least for rudder. As you pointed out, there is the 
1) cost aspect; we pay for rudder API calls as well as data ingestion - by "enabling the rudder development keys for local dev testing" does this open the door for devs for sending telemetry when they have no special reason to (as by default)? In general we should make it harder to do so, not easier. For local telemetry development there are other solutions - see https://docs.google.com/document/d/1YpBLoA-4rRu6n_H-2oGMHfOfzB9r7oMbjBHp5R07ZLs/edit (Option2). 
I would also add the
2) semantics - is the dev events semantics same as the one for the test events? I know over time there have been test-specific dashboards developed, sourcing test telemetry, they don't look especially used or maintained, but maybe @saturninoabril can weigh in if dev telemetry would skew existing metrics or not.

cc: @lieut-data

**Comment by lieut-data** [comment]
Thanks, @catalintomai -- based on your and Nick's feedback, I'll restore the `dev` default in which no Rudder keys are configured. Test behaviour and those less used dashboards is retained when explicitly configured.

**Review by isacikgoz — APPROVED** [review]
LGTM!

**Review by saturninoabril — APPROVED** [review]
Thanks Jesse, overall looks good to me.

- The docker images will be tackled separately once https://mattermost.atlassian.net/browse/CLD-5645 has started to work on.
- I don't think there's a better transition here than to wait for the change to take effect and subsequently address the test servers/automation accordingly.

@catalintomai, I don't know if we're talking the same test telemetry but the one I remember was the one set up by Alex D (from past Data team). If so, we don't have access to such data/analysis. Probably fine if not being looked at nowadays.

**Review by saturninoabril — APPROVED** [review]
Thanks Jesse, overall looks good to me.

- The docker images will be tackled separately once https://mattermost.atlassian.net/browse/CLD-5645 has started to work on.
- I don't think there's a better transition here than to wait for the change to take effect and subsequently address the test servers/automation accordingly.

@catalintomai, I don't know if we're talking the same test telemetry but the one I remember was the one set up by Alex D (from past Data team). If so, we don't have access to such data/analysis. Probably fine if not being looked at nowadays.

**Comment by phoinixgrr** [comment]
@lieut-data I have no objections to merging this, as long as the security team also is informed and approves of the decision to embed keys directly into the code.

cc: @esarafianou Could you take a look at this?

**Comment by esarafianou** [comment]
@lieut-data these Rudder and Stripe keys are the ones we had discussed in the "Immutable Build Artifacts" document, right?

**Review by esarafianou — APPROVED** [review]
LGTM!

**Pull request merged**
